mesos-master-detector/src/main/scala/com/mesosphere/mesos/MasterDetector.scala

```diff
@@ -0,0 +1,137 @@
+  def getMaster()(implicit ex: ExecutionContext): CompletionStage[URL]
+
+  /** @return whether the master string is valid. */
+  def isValid(): Boolean
+}
+
+object MasterDetector {
+
+  /**
+    * Constructs a master detector base on the master string passed.
+    *
+    * @param master The master string should be one of:
+    *               host:port
+    *               http://host:port
+    *               zk://host1:port1,host2:port2,.../path
+    *               zk://username:password@host1:port1,host2:port2,.../path
+    * @return A master detector.
+    */
+  def apply(master: String, metrics: Metrics): MasterDetector = {
+    if (master.startsWith("zk://")) {
+      Zookeeper(master, metrics)
+    } else {
+      Standalone(master)
+    }
+  }
+}
+
+case class Zookeeper(master: String, metrics: Metrics) extends MasterDetector with StrictLogging {
+  require(master.startsWith("zk://"), s"$master does not start with zk://")
+
+  case class ZkUrl(auth: Option[String], servers: String, path: String)
+
+  implicit val mesosInfoRead: Reads[Protos.MasterInfo] = (
+    (JsPath \ "hostname").read[String] ~
+      (JsPath \ "port").read[Int] ~
+      (JsPath \ "id").read[String] ~
+      (JsPath \ "ip").read[Int]
+  ) { (hostname, port, id, ip) =>
+    Protos.MasterInfo.newBuilder().setHostname(hostname).setPort(port).setId(id).setIp(ip).build()
+  }
+
+  override def isValid(): Boolean = Try(parse()).map(_ => true).getOrElse(false)
+
+  override def getMaster()(implicit ex: ExecutionContext): CompletionStage[URL] = {
+    val ZkUrl(auth, servers, path) = parse()
+
+    val client = {
+      val clientBuilder = CuratorFrameworkFactory.builder().connectString(servers).retryPolicy(new RetryOneTime(100))
+      auth.foreach(userPassword => clientBuilder.authorization("digest", userPassword.getBytes))
+      clientBuilder.build()
+    }
+    client.start()
+
+    if (!client.blockUntilConnected(
+        client.getZookeeperClient.getConnectionTimeoutMs,
+        java.util.concurrent.TimeUnit.MILLISECONDS)) {
+      throw new IllegalStateException("Failed to connect to Zookeeper. Will exit now.")
+    }
+
+    val clientSettings = AsyncCuratorBuilderSettings(createOptions = Set.empty, compressedData = false)
+    val factory: AsyncCuratorBuilderFactory = AsyncCuratorBuilderFactory(client, clientSettings)
+    val store: ZooKeeperPersistenceStore = new ZooKeeperPersistenceStore(metrics, factory, parallelism = 1)
+
+    async {
+      val children = await(store.children(path, false)).get
+      logger.info(s"Found children $children")
+      val leader = children.filter(_.startsWith("json.info")).min
+
+      val leaderPath = s"$path/$leader"
+      logger.info(s"Connecting to Zookeeper at $servers and fetching Mesos master from $leaderPath.")
+
+      val Node(_, bytes) = await(store.read(leaderPath)).get
+      logger.info(s"Mesos leader data: ${bytes.decodeString(StandardCharsets.UTF_8)}")
+      client.close()
```

Comment:
We should not keep a dangling zk connection in case something goes wrong. Wdyt about adding:
```
async {
...
}.recover{
      case t: Throwable =>
          logger.error("Failed to get mesos master leader node from ZK: ", t)
          client.close()
    }
```